README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# colorhex

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/colorhex)](https://CRAN.R-project.org/package=colorhex)
[![R-CMD-check](https://github.com/drmowinckels/colorhex/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/drmowinckels/colorhex/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The goal of colorhex is to create an interface to
[color-hex.com](https://www.color-hex.com/), a website with hexidecimal
colors and information about them.

It also has lots of user-made palettes that can be used and browsed.

## Installation

<!-- You can install the released version of colorhex from [CRAN](https://CRAN.R-project.org) with: -->
<!-- ``` r -->
<!-- install.packages("colorhex") -->
<!-- ``` -->

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("drmowinckels/colorhex", ref = "main")
```

## Example

### Single colors

``` r
library(colorhex)

x <- get_color("#470f0f")
x
#> # Color-hex: #470f0f
#> RGB: 71, 15, 15
#> HSL: 0.00, 0.65, 0.17
#> CMYK: 0.00, 0.79, 0.79   0.72
#> triadic: #0f470f, #0f0f47
#> complementary: #0f4747
#> used in 2 palettes
plot(x)
```

<img src="man/figures/README-example-1.png" width="100%" />

``` r
x <- get_popular_colors()
x
#>  [1] "#ff80ed" "#065535" "#000000" "#133337" "#ffc0cb" "#ffffff" "#ffe4e1"
#>  [8] "#008080" "#ff0000" "#e6e6fa" "#ffd700" "#00ffff" "#ffa500" "#0000ff"
#> [15] "#ff7373" "#c6e2ff" "#40e0d0" "#b0e0e6" "#d3ffce" "#f0f8ff" "#666666"
#> [22] "#faebd7" "#bada55" "#fa8072" "#003366" "#ffb6c1" "#c0c0c0" "#ffff00"
#> [29] "#800000" "#800080" "#c39797" "#00ff00" "#7fffd4" "#fff68f" "#eeeeee"
#> [36] "#cccccc" "#f08080" "#20b2aa" "#ffc3a0" "#333333" "#66cdaa" "#c0d6e4"
#> [43] "#ff6666" "#ff00ff" "#cbbeb5" "#ffdab9" "#468499" "#ff7f50" "#afeeee"
#> [50] "#b4eeb4" "#00ced1" "#008000" "#660066" "#f6546a" "#0e2f44" "#b6fcd5"
#> [57] "#990000" "#696969" "#f5f5f5" "#daa520" "#000080" "#6897bb" "#808080"
#> [64] "#f5f5dc" "#088da5" "#8b0000" "#8a2be2" "#81d8d0" "#ccff00" "#ffff66"
#> [71] "#ff4040" "#dddddd" "#2acaea" "#101010" "#0a75ad" "#ff1493" "#420420"
#> [78] "#66cccc" "#a0db8e" "#999999" "#794044" "#3399ff" "#cc0000" "#00ff7f"
scales::show_col(x)
```

<img src="man/figures/README-pop-cols-1.png" width="100%" />

### Palettes

``` r
latest <- get_latest_palettes()
plot(latest)
```

<img src="man/figures/README-latest-1.png" width="100%" />

``` r
popular <- get_popular_palettes()
plot(popular)
```

<img src="man/figures/README-popular-palettes-1.png" width="100%" />

### ggplot2 scales

``` r
library(ggplot2)

ggplot(mtcars, aes(mpg)) +
   geom_density(aes(fill = disp, group = disp)) +
   scale_fill_palettehex_c(popular)
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" />

``` r

ggplot(mtcars, aes(mpg)) +
  geom_density(aes(fill = disp, group = disp)) +
  scale_fill_palettehex_c(popular, 3)
```

<img src="man/figures/README-unnamed-chunk-2-2.png" width="100%" />

``` r

ggplot(mtcars, aes(mpg, disp, colour = factor(cyl))) +
   geom_point() +
   scale_color_palettehex_d(popular)
```

<img src="man/figures/README-unnamed-chunk-2-3.png" width="100%" />

``` r

ggplot(mtcars, aes(mpg, disp, colour = factor(cyl))) +
   geom_point() +
   scale_color_palettehex_d(popular, 1872)
```

<img src="man/figures/README-unnamed-chunk-2-4.png" width="100%" />

``` r
x <- get_color("#008080")

ggplot(mtcars, aes(mpg)) +
   geom_density(aes(fill = disp, group = disp)) +
   scale_fill_colorhex_c(x)
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="100%" />

``` r

ggplot(mtcars, aes(mpg)) +
  geom_density(aes(fill = disp, group = disp)) +
  scale_fill_colorhex_c(x, "tints")
```

<img src="man/figures/README-unnamed-chunk-3-2.png" width="100%" />

``` r

ggplot(mtcars, aes(mpg)) +
  geom_density(aes(fill = disp, group = disp)) +
  scale_fill_colorhex_c(x, "shades")
```

<img src="man/figures/README-unnamed-chunk-3-3.png" width="100%" />

``` r

ggplot(mtcars, aes(mpg, disp, colour = factor(cyl))) +
   geom_point() +
   scale_color_colorhex_d(x, "triadic")
```

<img src="man/figures/README-unnamed-chunk-3-4.png" width="100%" />

``` r

ggplot(mtcars, aes(mpg, disp, colour = factor(cyl))) +
   geom_point() +
   scale_color_colorhex_d(x, "shades")
```

<img src="man/figures/README-unnamed-chunk-3-5.png" width="100%" />


NEWS.md

# colorhex 0.1.4
- fixes failure of build if server is unresponsive

# colorhex 0.1.1
- cran submission

# colorhex 0.1.01

* With ggplot2 scales.

# colorhex 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* First release


R/api.R
query_colorhex <- function(){
  if(!curl::has_internet()){
    cli::cli_alert_warning("Not connected to internet.")
    return(invisible(NULL))
  }
  req <- httr2::request(colour_url())
  req <- httr2::req_retry(req,
                          backoff = ~ 10,
                          is_transient = ~ httr2::resp_status(.x) > 400)
  req <- httr2::req_error(req,
                   is_error = function(resp) FALSE,
                   body = error_body)
  req
}

colour_url <- function(full = TRUE){
  url <- "www.color-hex.com"
  if(!full)
    return(url)
  paste0("https://", url, "/")
}

error_body <- function(resp) {
  httr2::resp_body_json(resp)$error
}

status_ok <- function(req){
  test <- httr2::req_perform(req)
  if(httr2::resp_status(test) > 400 ){
    cli::cli_alert_warning("Cannot connect to service.")
    cli::cli_inform(httr2::resp_status_desc(test))
    return(FALSE)
  }
  TRUE
}


R/utils.R

#' Validate Hex
#'
#' validate if string is hexidecimal
#' color code
#'
#' @param x hexidecimal character
#' @return logical. TRUE if object is a hexidecimal code
#'
#' @export
is_hex <- function(x){
  grepl("^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$", x)
}


chartable <- function(table){
  x <- rvest::html_table(table)[[1]]
  as.character(x[-1, ])[-1]
}

strip_html <- function(x){
    rvest::html_text(
      rvest::read_html(x)
    )
}

get_nums <- function(x){
  as.numeric(strsplit(x, "\\D+")[[1]][-1])
}

get_pal_color <- function(x){
  x <- x[grepl("style", x)]
  get_bkg_color(x)
}

get_bkg_color <- function(x){
  x <- strsplit(x, "background-color:")
  x <- sapply(x, function(x) x[2])

  x <- gsub(';|\\\">|</div>| ', '', x)
  sapply(x, fix_hex)
}

fix_hex <- function(x){
  if(!is_hex(x)){
    cli::cli_abort("'{x}' is not a valid hexidecimal colour.")
  }
  indx <- ifelse(nchar(x) == 4, TRUE, FALSE)

  x[indx] <-  paste0(x[indx], gsub("#", "", x[indx]))
  x
}

nchar <- function(x){
  j <- strsplit(x, "")
  j <- lapply(j, length)
  unlist(j)
}

randcol <- function(){
  sample(1:255, 1)
}


R/palette.R
#' Get latest palettes
#'
#' Retrieve the most recently made palettes
#' from www.color-hex.com
#'
#' @return data.frame with name, id and colours
#' @export
#'
#' @examples
#' if(curl::has_internet()){
#' get_latest_palettes()
#' }
get_latest_palettes <- function(){
  req <- query_colorhex()
  if(is.null(req))
    return(invisible(NULL))
  req <- httr2::req_url_path_append(
    req, "color-palettes")
  if(!status_ok(req))
    return(invisible(NULL))
  resp <- httr2::req_perform(req)
  resp <- httr2::resp_body_html(resp)
  get_pals(resp)
}

#' Get most popular palettes
#'
#' Retrieve the palettes most users have
#' checked as favorites from www.color-hex.com
#'
#' @return data.frame with name, id and colours
#' @export
#'
#' @examples
#' if(curl::has_internet()){
#' get_popular_palettes()
#' }
get_popular_palettes <- function(){
  req <- query_colorhex()
  if(is.null(req))
    return(invisible(NULL))
  req <- httr2::req_url_path_append(
    req,
    "color-palettes",
    "popular.php")
  if(!status_ok(req))
    return(invisible(NULL))
  resp <- httr2::req_perform(req)
  resp <- httr2::resp_body_html(resp)
  get_pals(resp)
}

#' Get palettes from id
#'
#' Get palette information from www.color-hex.com
#' based on the palette id (can be found in the url)
#'
#' @param id numeric id of a palette
#'
#' @return data.frame with palette information
#' @export
#'
#' @examples
#' if(curl::has_internet()){
#' get_palette(103107)
#'
#' # Lookup multiple palettes
#' id <- c(103161, 103107)
#' get_palette(id)
#' }
get_palette <- function(id){
  x <- lapply(id, get_pal)
  do.call(rbind, x)
}

#' @export
plot.palettehex <- function(x, ...){
  colr <- apply(x, 1, unnest_pal)

  nrows <- max(sapply(colr, nrow))+1
  ncols <- length(colr)

  oldpar <- graphics::par(no.readonly = TRUE)
  on.exit(graphics::par(oldpar))
  graphics::par(mar = c(0, 0, 0, 0))
  graphics::plot.new()
  graphics::plot(c(-2, nrows-.2), c(.6, ncols+.3),
       type = "n", xlab = "", ylab = "",
       axes = FALSE
  )
  for(i in 1:length(colr)){
    tmp <- colr[[i]]
    graphics::text(1, i+.1,
                   sprintf("%s (%s)", tolower(tmp$name[[1]]), tmp$id[[1]]),
                   cex = .7, pos = 2)
    for(j in 1:nrow(tmp)){
      graphics::rect(j, i-.3, j+1, i+.3, col=tmp$hex[[j]], border = NA)
    }
  }

}

# helpers ----

get_pal <- function(id){
  req <- query_colorhex()
  if(is.null(req))
    return(invisible(NULL))
  req <- httr2::req_url_path_append(
    req,
    "color-palette",
    id)
  if(!status_ok(req))
    return(invisible(NULL))
  resp <- httr2::req_perform(req)
  resp <- httr2::resp_body_html(resp)

  tables <- rvest::html_nodes(resp, "table")
  tables <- rvest::html_table(tables[1], fill = TRUE)[[1]]

  palettehex(
    gsub(" Color Palette", "",
         rvest::html_text(rvest::html_nodes(resp, "h1"))),
    id,
    list(tables[,2])
  )
}

get_pals <- function(resp, class = "palettecontainerlist"){
  path <- paste0('//*[@class="',class, '"]')
  pal <- rvest::html_nodes(resp, xpath = path)
  pal2 <- as.character(pal)
  pal2 <- strsplit(pal2, "\n")

  title <- sapply(pal2, function(x) strip_html(x[length(x)-1]))
  number <- unlist(sapply(pal2, function(x) get_nums(x[2])[[1]]))

  palettehex(
    title,
    number,
    lapply(pal2, get_pal_color)
  )
}

palettehex <- function(name, id, palette){
  ret <- data.frame(
    name = name,
    id = id,
    stringsAsFactors = FALSE
  )
  ret$palette <- palette

  structure(
    ret,
    class = c("palettehex", "data.frame"))
}

unnest_pal <- function(x){
  k <- cbind(x[1], x[2], unname(unlist(x[3])))
  k <- as.data.frame(k, stringsAsFactors = FALSE)
  names(k) <- c("name", "id", "hex")
  row.names(k) <- NULL
  k$num <- 1:nrow(k)
  k
}



R/scale_colorhex.R
#' Colour scales for ggplot2
#'
#' Colour and fill scales for ggplot2
#' plots, using object of class \code{colorhex}
#' as basis for colour choices.
#'
#' The \code{colorhex} class is a list where there
#' is a variety of extra information on the hex
#' colour selected. This information can be used
#' to create colour scales to be used in ggplot2.
#'
#' @param x object of class \code{colorhex}
#' @param type character. Type of colours to use.
#'   One of c("complementary", "triadic" (default), "shades", "tints", "related")
#' @param reverse logical. If scale should b reversed (default: FALSE)
#' @param ... arguments to be passed to \code{\link[ggplot2]{discrete_scale}}
#'
#' @name scale-colorhex
#' @return a ggplot2-proto
#' @examples
#' if(curl::has_internet()){
#' library(ggplot2)
#'
#' x <- get_color("#008080")
#'
#' ggplot(mtcars, aes(mpg)) +
#'    geom_density(aes(fill = disp, group = disp)) +
#'    scale_fill_colorhex_c(x)
#'
#' ggplot(mtcars, aes(mpg)) +
#'   geom_density(aes(fill = disp, group = disp)) +
#'   scale_fill_colorhex_c(x, "tints")
#'
#' ggplot(mtcars, aes(mpg)) +
#'   geom_density(aes(fill = disp, group = disp)) +
#'   scale_fill_colorhex_c(x, "shades")
#'
#' ggplot(mtcars, aes(mpg, disp, colour = factor(cyl))) +
#'    geom_point() +
#'    scale_color_colorhex_d(x, "triadic")
#'
#' ggplot(mtcars, aes(mpg, disp, colour = factor(cyl))) +
#'    geom_point() +
#'    scale_color_colorhex_d(x, "shades")
#'  }
NULL
#> NULL

#' @describeIn scale-colorhex Discrete colour scale
#' @export
scale_colour_colorhex_d <- function(x,
                                    type = "triadic",
                                    reverse = FALSE,
                                    ...) {
  ggplot2::discrete_scale(
    "colour", "colorhex_d",
    colorhex_pal(x, type, reverse),
    ...
  )
}

#' @describeIn scale-colorhex Discrete colour scale
#' @export
scale_color_colorhex_d <- scale_colour_colorhex_d

#' @describeIn scale-colorhex Discrete fill scale
#' @export
scale_fill_colorhex_d <- function(x,
                                  type = "triadic",
                                  reverse = FALSE,
                                  ...) {
  ggplot2::discrete_scale(
    "fill", "colorhex_d",
    colorhex_pal(x, type, reverse),
    ...
  )
}

#' @describeIn scale-colorhex Continuous colour scale
#' @export
scale_colour_colorhex_c <- function(x,
                                    type = "complementary",
                                    reverse = FALSE,
                                    ...) {
  type <- match.arg(type, c("complementary", "triadic",
                            "shades", "tints", "related"))
  n <- length(c(x$hex, x[[type]]))
  cols <- colorhex_pal(x, type, reverse)(n)

  ggplot2::scale_colour_gradientn(colours = cols,
                                  ...)
}

#' @describeIn scale-colorhex Continuous colour scale
#' @export
scale_color_colorhex_c <- scale_colour_colorhex_c

#' @describeIn scale-colorhex Continuous fill scale
#' @export
scale_fill_colorhex_c <- function(x,
                                  type = "complementary",
                                  reverse = FALSE,
                                  ...) {
  type <- match.arg(type, c("complementary", "triadic",
                            "shades", "tints", "related"))
  n <- length(c(x$hex, x[[type]]))
  cols <- colorhex_pal(x, type, reverse)(n)

  ggplot2::scale_fill_gradientn(colours = cols,
                                ...)
}

# pal-maker ----
colorhex_pal <- function(x,
                         type = "complementary",
                         reverse = FALSE) {

  type <- match.arg(type, c("complementary", "triadic",
                            "shades", "tints", "related"))

  function(n) {
    cols <- c(x$hex, x[[type]])

    if (n > length(cols))
      warning(sprintf("colorhex only has %s colors.", length(cols)))

    cols <- cols[1:n]
    cols <- cols[!is.na(cols)]

    if (!reverse) cols else rev(cols)
  }
}


R/scale_palettehex.R
#' Colour scales for ggplot2
#'
#' Colour and fill scales for ggplot2
#' plots, using object of class \code{palettehex}.
#'
#' The \code{palettehex} class is a data.frame of
#' many palettes. This function takes such a data.frame
#' and a choice of palette by name, id or numeric index
#' can be made for the scale.
#'
#' @param x object of class \code{palettehex}
#' @param which selection of which palette of a set to choose. Either by name,
#'     id or numeric index.
#' @param reverse logical. If scale should b reversed (default: FALSE)
#' @param ... arguments to be passed to \code{\link[ggplot2]{discrete_scale}}
#'
#' @name scale-palettehex
#' @return ggplot2-proto
#' @examples
#' if(curl::has_internet()){
#' library(ggplot2)
#'
#' x <- get_popular_palettes()
#'
#' ggplot(mtcars, aes(mpg)) +
#'    geom_density(aes(fill = disp, group = disp)) +
#'    scale_fill_palettehex_c(x)
#'
#' ggplot(mtcars, aes(mpg)) +
#'   geom_density(aes(fill = disp, group = disp)) +
#'   scale_fill_palettehex_c(x, 3)
#'
#' ggplot(mtcars, aes(mpg, disp, colour = factor(cyl))) +
#'    geom_point() +
#'    scale_color_palettehex_d(x)
#'
#' ggplot(mtcars, aes(mpg, disp, colour = factor(cyl))) +
#'    geom_point() +
#'    scale_color_palettehex_d(x, 1872)
#' }
NULL
#> NULL

#' @describeIn scale-palettehex Discrete colour scale
#' @export
scale_colour_palettehex_d <- function(x,
                                    which = 1,
                                    reverse = FALSE,
                                    ...) {
  ggplot2::discrete_scale(
    "colour", "palettehex_d",
    palettehex_pal(x, which, reverse),
    ...
  )
}

#' @describeIn scale-palettehex Discrete colour scale
#' @export
scale_color_palettehex_d <- scale_colour_palettehex_d

#' @describeIn scale-palettehex Discrete fill scale
#' @export
scale_fill_palettehex_d <- function(x,
                                  which = 1,
                                  reverse = FALSE,
                                  ...) {
  ggplot2::discrete_scale(
    "fill", "palettehex_d",
    palettehex_pal(x, which, reverse),
    ...
  )
}

#' @describeIn scale-palettehex Continuous colour scale
#' @export
scale_colour_palettehex_c <- function(x,
                                      which = 1,
                                      reverse = FALSE,
                                    ...) {
  n <- length(find_pal(x, which))
  cols <- palettehex_pal(x, which, reverse)(n)

  ggplot2::scale_colour_gradientn(colours = cols,
                                  ...)
}

#' @describeIn scale-palettehex Continuous colour scale
#' @export
scale_color_palettehex_c <- scale_colour_palettehex_c

#' @describeIn scale-palettehex Continuous fill scale
#' @export
scale_fill_palettehex_c <- function(x,
                                    which = 1,
                                  reverse = FALSE,
                                  ...) {
  n <- length(find_pal(x, which))
  cols <- palettehex_pal(x, which, reverse)(n)

  ggplot2::scale_fill_gradientn(colours = cols,
                                ...)
}

# pal-maker ----
palettehex_pal <- function(x,
                         which = 1,
                         reverse = FALSE) {
  cols <- find_pal(x, which)

  function(n) {
    if (n > length(cols))
      warning(sprintf("palettehex only has %s colors.", length(cols)))

    cols <- cols[1:n]
    cols <- cols[!is.na(cols)]

    if (!reverse) cols else rev(cols)
  }
}


find_pal <- function(x, which){

  idx <- match(which, x$name)
  if(!is.na(idx)) return(x$palette[[idx]])

  idx <- match(which, x$id)
  if(!is.na(idx)) return(x$palette[[idx]])

  x$palette[[which]]

}


inst/js/color-palette.js
// js API connection from https://github.com/airtoxin/color-hex/

import request from "request-promise";
import cheerio from "cheerio";

export const listPopular = () => {
    return request("http://www.color-hex.com/color-palettes/popular.php")
        .then((body) => {
            const $ = cheerio.load(body);
            const $palettes = $(".palettecontainerlist");
            return $palettes.map((i, pel) => {
                const $palette = $(pel);
                const name = $palette.text().trim();
                const colors = $palette.find(".palettecolordiv").map((i, cel) => {
                    return $(cel).css("background-color");
                }).get();
                const id = $palette.find("a").attr("href").match(/.*color-palette\/(\d+)/)[1];
                return {
                    id,
                    name,
                    colors
                };
            }).get();
        });
};


R/color.R
#' Get popular colour
#'
#' www.color-hex.com has a list of colours
#' that have been liked by the most users.
#' This function will retrieve all of these.
#'
#' @return character vector of hex colours
#' @export
#'
#' @examples
#' if(curl::has_internet()){
#' get_popular_colors()
#' }
get_popular_colors <- function(){
  req <- httr2::request(colour_url())
  if(is.null(req))
    return(invisible(NULL))
  req <- httr2::req_url_path_append(
    req,
    "popular-colors.php")
  if(!status_ok(req))
    return(invisible(NULL))

  resp <- httr2::req_perform(req)
  resp <- httr2::resp_body_html(resp)
  cols <- rvest::html_nodes(
    resp,
    xpath = '//*[@class="colordva"]')
  cols <- as.character(cols)
  get_bkg_color(cols)
}

#' Generate random HEX colour
#'
#' @return character hex value
#' @export
#'
#' @examples
#' get_random_color()
get_random_color <- function(){
  grDevices::rgb(randcol(),
                 randcol(),
                 randcol(),
                 maxColorValue = 255)
}

#' Get color information
#'
#' Get color information from www.color-hex.com
#' of a hex-color.
#'
#' @param hex character string that is a hexidecimal color
#'
#' @return list of class 'colorhex'
#' @export
#'
#' @examples
#' if(curl::has_internet()){
#' get_color("#470f0f")
#' }
get_color <- function(hex){
  hex <- fix_hex(hex)
  req <- query_colorhex()
  if(is.null(req))
    return(invisible(NULL))

  req <- httr2::req_url_path_append(
    req,
    "color",
    gsub("^#", "", hex))

  if(!status_ok(req))
    return(invisible(NULL))

  resp <- httr2::req_perform(req)
  resp <- httr2::resp_body_html(resp)
  tables <- rvest::html_nodes(resp, "table")
  tables <- lapply(tables, rvest::html_table, fill = TRUE)
  prim <- as.data.frame(t(tables[[1]]))
  names(prim) <- as.character(unlist(prim[1,]))
  row.names(prim) <- NULL
  prim <- prim[-1,]

  rows <- rvest::html_nodes(resp,
                            xpath = '//*[@class="colordvconline"]')
  rows <- rvest::html_text(rows)
  rows <- gsub(" \n", "", rows)
  rows <- sapply(rows, fix_hex)

  ret <- list(
    hex = hex,
    space = prim,
    base = tables[[2]],
    triadic = NA_character_,
    analogous = NA_character_,
    complementary = NA_character_,
    shades = rows[1:11],
    tints = rows[12:22],
    related = rows[22:length(rows)],
    palettes = get_pals(resp, "palettecontainerlist narrow")
  )

  if(length(tables) > 2){
    ex <- lapply(3:5, function(x){
      j <- unique(unlist(tables[[x]]))
      sapply(j[j!=""], fix_hex)
    })
    ret$triadic = ex[[1]]
    ret$analogous = ex[[2]]
    ret$complementary = ex[[3]]
  }

  colorhex(ret)
}

colorhex <- function(x){
  stopifnot(names(x) %in% c("hex",
                            "space", "base",
                            "complementary", "analogous",
                            "triadic", "shades", "tints",
                            "related", "palettes"))

  structure(
    x,
    class = "colorhex"
  )
}

#' @export
format.colorhex <- function(x, ...){
  c(
    sprintf("# Color-hex: %s", x$hex),
    sprintf("RGB: %s", paste0(x$space[, "RGB"], collapse=", ")),
    sprintf("HSL: %s", paste0(x$space[, "HSL"], collapse=", ")),
    sprintf("CMYK: %s", paste0(x$space[, "CMYK"], collapse=", ")),
    sprintf("triadic: %s", paste0(x$triadic, collapse = ", ")),
    sprintf("complementary: %s", x$complementary),
    sprintf("used in %s palettes", nrow(x$palettes))
  )
}

#' @export
print.colorhex <- function(x, ...){
  cat(format(x), sep="\n")
  invisible(x)
}

#' @export
plot.colorhex <- function(x,
                          type = c("complementary", "triadic",
                                   "analogous", "shades", "tints",
                                   "related"),
                          labels = TRUE, ...){

  type <- match.arg(type,
                    c("complementary", "triadic",
                      "analogous", "shades", "tints", "related"),
                    several.ok = TRUE)

  x <- lapply(type, function(y) if(y != "hex") c(x$hex, x[[y]]) else x[[y]])
  names(x) <- type

  ncols <- length(type)
  nrows <- max(sapply(x, length))+.5

  oldpar <- graphics::par(no.readonly = TRUE)
  on.exit(graphics::par(oldpar))
  graphics::par(mar = c(0, 0, 0, 0))
  graphics::plot.new()
  graphics::plot(c(-.1, nrows+.3), c(.5, ncols+.5),
                 type = "n", xlab = "", ylab = "",
                 axes = FALSE
  )

  for(i in 1:length(type)){
    tmp <- x[[type[i]]]
    graphics::text(1, i, type[i], cex = 1, pos = 2)
    for(j in 1:length(tmp)){
      graphics::rect(j, i-.4, j+1, i+.4, col=tmp[j], border = NA)
      if(labels){
        graphics::rect(j+.2, i-.1, j+.8, i+.1, col="white", border = NA)
        graphics::text(j+.5, i, tmp[j], cex = .7)
      }
    }
  }

}
